service/forward_service.cc

```diff
@@ -0,0 +1,176 @@
+#include "service/forward_service.hh"
+
+#include <boost/range/algorithm/remove_if.hpp>
+#include <seastar/core/future-util.hh>
+#include <seastar/core/coroutine.hh>
+
+#include "schema_registry.hh"
+#include "database.hh"
+#include "query-ranges-to-vnodes.hh"
+#include "locator/abstract_replication_strategy.hh"
+#include "gms/gossiper.hh"
+#include "log.hh"
+
+#include "service/pager/query_pagers.hh"
+#include "cql3/selection/selectable.hh"
+#include "cql3/selection/selectable-expr.hh"
+#include "cql3/selection/selection.hh"
+#include "cql3/selection/raw_selector.hh"
+#include "cql3/cql_config.hh"
+#include "cql3/column_identifier.hh"
+#include "cql3/query_options.hh"
+#include "cql3/result_set.hh"
+#include "cql_serialization_format.hh"
+
+namespace service {
+
+static logging::logger flogger("forward_service");
+
+static inline
+const dht::token& end_token(const dht::partition_range& r) {
+    static const dht::token max_token = dht::maximum_token();
+    return r.end() ? r.end()->value().token() : max_token;
+}
+
+static inet_address_vector_replica_set get_live_endpoints(keyspace& ks, const dht::token& token) {
+    auto erm = ks.get_effective_replication_map();
+    inet_address_vector_replica_set eps = erm->get_natural_endpoints_without_node_being_replaced(token);
+    auto itend = boost::range::remove_if(eps, std::not1(std::bind1st(std::mem_fn(&gms::gossiper::is_alive), &gms::get_local_gossiper())));
+    eps.erase(itend, eps.end());
+    return eps;
+}
+
+locator::token_metadata_ptr forward_service::get_token_metadata_ptr() const noexcept {
+    return _shared_token_metadata.get();
+}
+
+future<> forward_service::stop() {
+    return make_ready_future<>();
+}
+
+future<query::forward_result> forward_service::execute(query::forward_request req) {
+    schema_ptr schema = local_schema_registry().get(req.cmd.schema_version);
+    auto command = make_lw_shared<query::read_command>(std::move(req.cmd));
+
+    flogger.debug("forward_request: received req={} from addr={}", req, _messaging.listen_address());
+
+    // create mocked selection
+    auto selection = [&]{
+        switch (req.type) {
+            case query::forward_request::reduction_type::count: {
+                auto selectable = cql3::selection::make_count_rows_function_expression();
+                auto column_identifier = make_shared<cql3::column_identifier>("count", false);
+                std::vector<shared_ptr<cql3::selection::raw_selector>> raw_selectors;
+                raw_selectors.emplace_back(make_shared<cql3::selection::raw_selector>(selectable, column_identifier));
+                return cql3::selection::selection::from_selectors(_db.local(), schema, std::move(raw_selectors));
```

Comment:
Okay, now I see why. We don't have aggregation outside cql3 and this is how we reuse the aggregation code.

It's super sad but I won't ask you to refactor the aggregation code to be language neutral just for this.

Reply:
(at least it's just an implementation dependency, not an interface dependency).